### ISSUE TITLE
LLVM Assertion `FragmentOffset >= OffsetInBits && "overlapping or duplicate fragments"' failed

### ISSUE THREAD

**Issue by andrewrk** [ziglang/zig#6408]
To reproduce, do this with an asserts-enabled LLVM build:

```
./zig test ../lib/std/std.zig --release-fast
```

[Upstream bug](https://bugs.llvm.org/show_bug.cgi?id=34435)

We weren't hitting this in zig 0.6.0 - something happened in master branch since then that has caused us to now trigger this llvm bug.

**Comment by FireFox317** [comment]
My guess would be that this is caused by the fact that when we test the standard library, we put all the test code into a single executable. LLVM then tries to add debug information to this big file and causing this assertion to happen. When I ran all the zig files in the std folder separately, the assertion was not hit.

**Comment by FireFox317** [comment]
Building llvm11 with `zig cc/c++` sanitizers enabled, found an illegal instruction. This is also in the DwarfExpression code, so I have a feeling it is related to this issue.
```gdb
Program received signal SIGILL, Illegal instruction.
0x063a3680 in llvm::DwarfExpression::addExpression(llvm::DIExpressionCursor&&, unsigned int) ()
(gdb) bt
#0  0x063a3680 in llvm::DwarfExpression::addExpression(llvm::DIExpressionCursor&&, unsigned int) ()
#1  0x06364eac in llvm::DwarfDebug::emitDebugLocValue(llvm::AsmPrinter const&, llvm::DIBasicType const*bgValueLoc const&, llvm::DwarfExpression&) ()
#2  0x063616b4 in llvm::DebugLocEntry::finalize(llvm::AsmPrinter const&, llvm::DebugLocStream::ListBuilm::DIBasicType const*, llvm::DwarfCompileUnit&) ()
#3  0x06360d30 in llvm::DwarfDebug::collectEntityInfo(llvm::DwarfCompileUnit&, llvm::DISubprogram constDenseSet<std::__1::pair<llvm::DINode const*, llvm::DILocation const*>, llvm::DenseMapInfo<std::__1::paiINode const*, llvm::DILocation const*> > >&) ()
#4  0x06362a38 in llvm::DwarfDebug::endFunctionImpl(llvm::MachineFunction const*) ()
#5  0x06347ea4 in llvm::DebugHandlerBase::endFunction(llvm::MachineFunction const*) ()
#6  0x0632faf0 in llvm::AsmPrinter::emitFunctionBody() ()
#7  0x054ecc50 in llvm::ARMAsmPrinter::runOnMachineFunction(llvm::MachineFunction&) ()
#8  0x065cfe3c in llvm::MachineFunctionPass::runOnFunction(llvm::Function&) ()
#9  0x07350370 in llvm::FPPassManager::runOnFunction(llvm::Function&) ()
#10 0x073589c0 in llvm::FPPassManager::runOnModule(llvm::Module&) ()
#11 0x07350ba8 in llvm::legacy::PassManagerImpl::run(llvm::Module&) ()
#12 0x020423dc in ZigLLVMTargetMachineEmitToFile ()
#13 0x01ebfa5c in codegen_build_object(CodeGen*) ()
#14 0x01eb46b0 in zig_stage1_build_object ()
#15 0x01c9a850 in Compilation.performAllTheWork ()
#16 0x01c7e3c8 in Compilation.update ()
#17 0x01d59e30 in Compilation.updateSubCompilation ()
#18 0x01cbf56c in Compilation.buildStaticLibFromZig ()
#19 0x01c8aa5c in Compilation.performAllTheWork ()
#20 0x01c7e3c8 in Compilation.update ()
#21 0x01c57678 in main.updateModule ()
#22 0x01bfaa38 in main.buildOutputType ()
#23 0x01bcc2f4 in main ()
```

**Comment by mikdusan** [comment]
(on llvm12 branch, macos x86_64, debug llvm, debug zig)

As I was reducing the case, llvm assert no longer manifested but a zig test failure appeared. Normally I'd say it was a different bug but I haven't ruled out and thinking still on the same bug track:

```sh
[llvm12⋅2-2] tiberius▸ zig test -OReleaseFast lib/std/fs.zig

Test [42/42] fs.path.test "extension"...
====== expected this output: =========
␃

======== instead found this: =========
.␃

======================================
First difference occurs on line 1:
expected:

^
found:
.
^
thread 816183 panic: test failure
error: the following test command crashed:
lib/std/zig-cache/o/8b1a4a2a2bd874e772cf2b43e1de938d/test /Users/mike/project/zig/work/llvm12/_build/zig
```

so what strikes me as strange is this inspection from lldb. First, here is the function I am concerned with:

```zig
fn testExtension(path: []const u8, expected: []const u8) void {
    std.testing.expectEqualStrings(expected, extension(path));
}
```

Now here are 2 lldb backtraces concerning `fs.path.testExtension`. First from a `-OReleaseFast` binary and second from  a debug binary:

```diff
!     frame #7: 0x0000000100011399 test`fs.path.test "extension" [inlined] fs.path.testExtension(expected=(ptr = "", len = 0)) at path.zig:1219:35 [opt]
      frame #8: 0x0000000100011377 test`fs.path.test "extension" at path.zig:1224 [opt]
      frame #9: 0x0000000100014d70 test`std.start.main at test_runner.zig:69:28 [opt]
```

```diff
!     frame #13: 0x0000000100071d29 test`fs.path.testExtension(path=<unavailable>, expected=(ptr = "", len = 0)) at path.zig:1220:1
      frame #14: 0x000000010000bc8b test`fs.path.test "extension" at path.zig:1224:18
      frame #15: 0x000000010000e654 test`std.special.main at test_runner.zig:69:28
```

---

So for both release and debug builds, the first argument has some kind of issue. In release it's actually not even present but I am not sure if this has to do with inlining? In debug it says `<unavailable>` which smells like a bad sign.

Might we have bad debug generation? I'll try to get a real reduction but wanted to record current findings.

**Comment by LemonBoy** [comment]
Shit is hitting the fan at high speed. Calling `extension(".")` is triggering some miscompilation (or it's just UB on our side) in the LLVM optimization pipelines.
The missing argument is a side-effect of the optimizations kicking in.

Here's a small reproducer:
```zig
pub export fn _start() void {
    @setAlignStack(16);
    _ = extension(".");
}

const builtin = @import("builtin");
pub fn panic(msg: []const u8, error_return_trace: ?*builtin.StackTrace) noreturn {
    while (true) {}
}

fn eql(comptime T: type, a: []const T, b: []const T) bool {
    @setRuntimeSafety(false);
    if (a.len != b.len) return false;
    if (a.ptr == b.ptr) return true;
    for (a) |item, index| {
        if (b[index] != item) return false;
    }
    return true;
}

fn lastIndexOfLinear(comptime T: type, haystack: []const T, needle: []const T) usize {
    @setRuntimeSafety(false);
    var i: usize = haystack.len -% needle.len;
    while (true) : (i -%= 1) {
        if (eql(T, haystack[i .. i +% needle.len], needle)) return i;
        if (i == 0) return 0;
    }
}

fn extension(path: []const u8) []const u8 {
    @setRuntimeSafety(false);
    const filename = path;
    const index = lastIndexOfLinear(u8, filename, ".");
    if (index == 0) return "XXX";
    @breakpoint();
    return path[index..];
}
```

Running it in `ReleaseFast` mode shows `@breakpoint` being triggered even if `index = 0`.

Reduced LLVM IR:

```llvm
; ModuleID = 'foo'
source_filename = "foo"
target datalayout = "e-m:e-p270:32:32-p271:32:32-p272:64:64-i64:64-f80:128-n8:16:32:64-S128"
target triple = "x86_64-unknown-linux-gnu"

%"[]u8" = type { i8*, i64 }
%std.builtin.StackTrace = type { i64, %"[]usize" }
%"[]usize" = type { i64*, i64 }

@0 = internal unnamed_addr constant [10 x i8] c"ivybridge\00", align 1
@1 = internal unnamed_addr constant [7 x i8] c"_start\00", align 1
@2 = internal unnamed_addr constant [2 x i8] c".\00", align 1
@3 = internal unnamed_addr constant %"[]u8" { i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i64 0, i64 0), i64 1 }, align 8
@4 = internal unnamed_addr constant %"[]u8" { i8* getelementptr inbounds ([2 x i8], [2 x i8]* @2, i64 0, i64 0), i64 1 }, align 8
@5 = internal unnamed_addr constant [4 x i8] c"XXX\00", align 1
@6 = internal unnamed_addr constant %"[]u8" { i8* getelementptr inbounds ([4 x i8], [4 x i8]* @5, i64 0, i64 0), i64 3 }, align 8

; Function Attrs: nobuiltin noreturn nounwind
define internal fastcc void @panic(%"[]u8"* nonnull readonly align 8 %0, %std.builtin.StackTrace* align 8 %1) unnamed_addr  {
Entry:
  %error_return_trace = alloca %std.builtin.StackTrace*, align 8
  store %std.builtin.StackTrace* %1, %std.builtin.StackTrace** %error_return_trace, align 8
  br label %WhileCond

WhileCond:                                        ; preds = %WhileCond, %Entry
  br label %WhileCond
}

; Function Attrs: nobuiltin noinline nounwind alignstack(16)
define void @_start()  {
Entry:
  %0 = alloca %"[]u8", align 8
  call fastcc void @extension(%"[]u8"* %0, %"[]u8"* @3)
  ret void
}

; Function Attrs: nobuiltin nounwind
define internal fastcc void @extension(%"[]u8"* nonnull sret(%"[]u8") %0, %"[]u8"* nonnull readonly align 8 %1) unnamed_addr  {
Entry:
  %filename = alloca %"[]u8", align 8
  %2 = alloca %"[]u8", align 8
  %index = alloca i64, align 8
  %3 = bitcast %"[]u8"* %1 to i8*
  %4 = bitcast %"[]u8"* %filename to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 8 %4, i8* align 8 %3, i64 16, i1 false)
  %5 = bitcast %"[]u8"* %filename to i8*
  %6 = bitcast %"[]u8"* %2 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 8 %6, i8* align 8 %5, i64 16, i1 false)
  %7 = call fastcc i64 @lastIndexOfLinear(%"[]u8"* %filename, %"[]u8"* @4)
  store i64 %7, i64* %index, align 8
  %8 = load i64, i64* %index, align 8
  %9 = icmp eq i64 %8, 0
  br i1 %9, label %Then, label %Else

Then:                                             ; preds = %Entry
  %10 = bitcast %"[]u8"* %0 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 8 %10, i8* align 8 bitcast (%"[]u8"* @6 to i8*), i64 16, i1 false)
  ret void

Else:                                             ; preds = %Entry
  br label %EndIf

EndIf:                                            ; preds = %Else
  call void @llvm.debugtrap()
  %11 = load i64, i64* %index, align 8
  %12 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 1
  %13 = load i64, i64* %12, align 8
  %14 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 0
  %15 = load i8*, i8** %14, align 8
  %16 = getelementptr inbounds i8, i8* %15, i64 %11
  %17 = sub nuw i64 %13, %11
  %18 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 0
  store i8* %16, i8** %18, align 8
  %19 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 1
  store i64 %17, i64* %19, align 8
  ret void
}

; Function Attrs: argmemonly nofree nosync nounwind willreturn
declare void @llvm.memcpy.p0i8.p0i8.i64(i8* noalias nocapture writeonly, i8* noalias nocapture readonly, i64, i1 immarg) 

; Function Attrs: nobuiltin nounwind
define internal fastcc i64 @lastIndexOfLinear(%"[]u8"* nonnull readonly align 8 %0, %"[]u8"* nonnull readonly align 8 %1) unnamed_addr  {
Entry:
  %result = alloca i64, align 8
  %i = alloca i64, align 8
  %2 = alloca %"[]u8", align 8
  %3 = alloca %"[]u8", align 8
  %4 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 1
  %5 = load i64, i64* %4, align 8
  %6 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 1
  %7 = load i64, i64* %6, align 8
  %8 = sub i64 %5, %7
  store i64 %8, i64* %i, align 8
  br label %WhileCond

WhileCond:                                        ; preds = %EndIf3, %Entry
  br label %WhileBody

WhileBody:                                        ; preds = %WhileCond
  %9 = load i64, i64* %i, align 8
  %10 = load i64, i64* %i, align 8
  %11 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 1
  %12 = load i64, i64* %11, align 8
  %13 = add i64 %10, %12
  %14 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 0
  %15 = load i8*, i8** %14, align 8
  %16 = getelementptr inbounds i8, i8* %15, i64 %9
  %17 = sub nuw i64 %13, %9
  %18 = getelementptr inbounds %"[]u8", %"[]u8"* %2, i32 0, i32 0
  store i8* %16, i8** %18, align 8
  %19 = getelementptr inbounds %"[]u8", %"[]u8"* %2, i32 0, i32 1
  store i64 %17, i64* %19, align 8
  %20 = bitcast %"[]u8"* %1 to i8*
  %21 = bitcast %"[]u8"* %3 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 8 %21, i8* align 8 %20, i64 16, i1 false)
  %22 = call fastcc i1 @eql(%"[]u8"* %2, %"[]u8"* %1)
  br i1 %22, label %Then, label %Else

Then:                                             ; preds = %WhileBody
  %23 = load i64, i64* %i, align 8
  store i64 %23, i64* %result, align 8
  %24 = load i64, i64* %result, align 8
  ret i64 %24

Else:                                             ; preds = %WhileBody
  br label %EndIf

EndIf:                                            ; preds = %Else
  %25 = load i64, i64* %i, align 8
  %26 = icmp eq i64 %25, 0
  br i1 %26, label %Then1, label %Else2

Then1:                                            ; preds = %EndIf
  store i64 0, i64* %result, align 8
  %27 = load i64, i64* %result, align 8
  ret i64 %27

Else2:                                            ; preds = %EndIf
  br label %EndIf3

EndIf3:                                           ; preds = %Else2
  %28 = load i64, i64* %i, align 8
  %29 = sub i64 %28, 1
  store i64 %29, i64* %i, align 8
  br label %WhileCond
}

; Function Attrs: nounwind
declare void @llvm.debugtrap() 

; Function Attrs: nobuiltin nounwind
define internal fastcc i1 @eql(%"[]u8"* nonnull readonly align 8 %0, %"[]u8"* nonnull readonly align 8 %1) unnamed_addr  {
Entry:
  %result = alloca i1, align 1
  %index = alloca i64, align 8
  %item = alloca i8, align 1
  %2 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 1
  %3 = load i64, i64* %2, align 8
  %4 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 1
  %5 = load i64, i64* %4, align 8
  %6 = icmp ne i64 %3, %5
  br i1 %6, label %Then, label %Else

Then:                                             ; preds = %Entry
  store i1 false, i1* %result, align 1
  %7 = load i1, i1* %result, align 1
  ret i1 %7

Else:                                             ; preds = %Entry
  br label %EndIf

EndIf:                                            ; preds = %Else
  %8 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 0
  %9 = load i8*, i8** %8, align 8
  %10 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 0
  %11 = load i8*, i8** %10, align 8
  %12 = icmp eq i8* %9, %11
  br i1 %12, label %Then1, label %Else2

Then1:                                            ; preds = %EndIf
  store i1 true, i1* %result, align 1
  %13 = load i1, i1* %result, align 1
  ret i1 %13

Else2:                                            ; preds = %EndIf
  br label %EndIf3

EndIf3:                                           ; preds = %Else2
  store i64 0, i64* %index, align 8
  %14 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 1
  %15 = load i64, i64* %14, align 8
  br label %ForCond

ForCond:                                          ; preds = %EndIf6, %EndIf3
  %16 = load i64, i64* %index, align 8
  %17 = icmp ult i64 %16, %15
  br i1 %17, label %ForBody, label %ForEnd

ForBody:                                          ; preds = %ForCond
  %18 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i32 0, i32 0
  %19 = load i8*, i8** %18, align 8
  %20 = getelementptr inbounds i8, i8* %19, i64 %16
  %21 = load i8, i8* %20, align 1
  store i8 %21, i8* %item, align 1
  %22 = load i64, i64* %index, align 8
  %23 = getelementptr inbounds %"[]u8", %"[]u8"* %1, i32 0, i32 0
  %24 = load i8*, i8** %23, align 8
  %25 = getelementptr inbounds i8, i8* %24, i64 %22
  %26 = load i8, i8* %25, align 1
  %27 = load i8, i8* %item, align 1
  %28 = icmp ne i8 %26, %27
  br i1 %28, label %Then4, label %Else5

Then4:                                            ; preds = %ForBody
  store i1 false, i1* %result, align 1
  %29 = load i1, i1* %result, align 1
  ret i1 %29

Else5:                                            ; preds = %ForBody
  br label %EndIf6

EndIf6:                                           ; preds = %Else5
  %30 = add nuw i64 %16, 1
  store i64 %30, i64* %index, align 8
  br label %ForCond

ForEnd:                                           ; preds = %ForCond
  store i1 true, i1* %result, align 1
  %31 = load i1, i1* %result, align 1
  ret i1 %31
}
```

**Comment by LemonBoy** [comment]
Bisected to [this LLVM patch](https://reviews.llvm.org/D93820). The GEPs in the starting LLVM IR look correct to me, perhaps it's because of the interaction with some other optimization pass.

CC @nikic

**Comment by nikic** [comment]
```
%"[]u8" = type { i8*, i64 }
%std.builtin.StackTrace = type { i64, %"[]usize" }
%"[]usize" = type { i64*, i64 }

@0 = internal unnamed_addr constant [2 x i8] c".\00", align 1
@1 = internal unnamed_addr constant %"[]u8" { i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i64 0, i64 0), i64 1 }, align 8

; Function Attrs: nounwind
define internal fastcc void @extension(%"[]u8"* nocapture nonnull sret(%"[]u8") %0) unnamed_addr #0 {
Entry:
  br label %EndIf.i.i

EndIf.i.i:                                        ; preds = %eql.exit.i, %Entry
  %i.0.in.i = phi i64 [ 1, %Entry ], [ %i.0.i, %eql.exit.i ]
  %i.0.i = add i64 %i.0.in.i, -1 
  %1 = getelementptr inbounds i8, i8* getelementptr inbounds ([2 x i8], [2 x i8]* @0, i64 0, i64 0), i64 %i.0.i
  %2 = icmp eq i8* %1, getelementptr inbounds ([2 x i8], [2 x i8]* @0, i64 0, i64 0) 
  br i1 %2, label %lastIndexOfLinear.exit, label %eql.exit.i

eql.exit.i:                                       ; preds = %EndIf.i.i
  %3 = load i8, i8* %1, align 1
  %.not1.i.i = icmp eq i8 %3, 46
  %4 = icmp eq i64 %i.0.i, 0
  %or.cond.i = or i1 %4, %.not1.i.i
  br i1 %or.cond.i, label %lastIndexOfLinear.exit, label %EndIf.i.i

lastIndexOfLinear.exit:                           ; preds = %EndIf.i.i, %eql.exit.i
  %i.0.lcssa.i = phi i64 [ ptrtoint (i8* getelementptr ([2 x i8], [2 x i8]* @0, i64 0, i64 sub (i64 0, i64 ptrtoint ([2 x i8]* @0 to i64))) to i64), %EndIf.i.i ], [ %i.0.i, %eql.exit.i ]
  %5 = icmp eq i64 %i.0.lcssa.i, 0
  br i1 %5, label %Then, label %EndIf

Then:                                             ; preds = %lastIndexOfLinear.exit
  %6 = bitcast %"[]u8"* %0 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* noundef nonnull align 8 dereferenceable(16) %6, i8* noundef nonnull align 8 dereferenceable(16) bitcast (%"[]u8"* @1 to i8*), i64 16, i1 false)
  ret void 

EndIf:                                            ; preds = %lastIndexOfLinear.exit
  call void @llvm.debugtrap() 
  %7 = getelementptr inbounds [2 x i8], [2 x i8]* @0, i64 0, i64 %i.0.lcssa.i
  %8 = sub nuw i64 1, %i.0.lcssa.i
  %9 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i64 0, i32 0
  store i8* %7, i8** %9, align 8
  %10 = getelementptr inbounds %"[]u8", %"[]u8"* %0, i64 0, i32 1
  store i64 %8, i64* %10, align 8
  ret void
} 
  
declare void @llvm.debugtrap()
  
declare void @llvm.memcpy.p0i8.p0i8.i64(i8* noalias nocapture writeonly, i8* noalias nocapture readonly, i64, i1 immarg)
```

After the referenced patch we no longer fold `ptrtoint (i8* getelementptr ([2 x i8], [2 x i8]* @0, i64 0, i64 sub (i64 0, i64 ptrtoint ([2 x i8]* @0 to i64))) to i64)` to `0`, and then `-jump-threading` on the above code folds the branch to `%EndIf` rather than `%Then`.

**Comment by LemonBoy** [comment]
That I understand, but given the difference is not folded to zero anymore I'd expect the comparison `%5` not to be folded at all.

**Comment by nikic** [comment]
Reduced:

```
RUN: opt -instsimplify -S < %s
@g = internal unnamed_addr constant [2 x i8] c".\00", align 1

define i1 @test() {
  %cmp = icmp eq i64 ptrtoint (i8* getelementptr ([2 x i8], [2 x i8]* @g, i64 0, i64 sub (i64 0, i64 ptrtoint ([2 x i8]* @g to i64))) to i64), 0
  ret i1 %cmp
}
```

This fold to `i1 false`, while it should be `i1 true` or not folded at all.

**Comment by nikic** [comment]
Slightly more reduced:

```
@g = internal unnamed_addr constant [2 x i8] c".\00", align 1

define i1 @test() {
  %cmp = icmp eq i8* getelementptr ([2 x i8], [2 x i8]* @g, i64 0, i64 sub (i64 0, i64 ptrtoint ([2 x i8]* @g to i64))), null
  ret i1 %cmp
}
```

By inspection, I believe this is the problematic fold: https://github.com/llvm/llvm-project/blob/5de09ef02e24d234d9fc0cd1c6dfe18a1bb784b0/llvm/lib/IR/ConstantFold.cpp#L1846-L1848 This is only valid for inbounds GEP.

**Comment by mikdusan** [comment]
@LemonBoy, thanks for the `llvm-reduce` help:

.ll is here → [_optimized.zip](https://github.com/ziglang/zig/files/6060374/_optimized.zip)

#### reproduced on macos with LLVM release, enable-asserts:
```
[llvm12⋅0-0] tiberius▸ llc _optimized.ll

Assertion failed: ((FrameIndexExprs.size() == 1 || llvm::all_of(FrameIndexExprs, [](FrameIndexExpr &FIE) { return FIE.Expr && FIE.Expr->isFragment(); })) && "conflicting locations for variable"), function addMMIEntry, file /Users/mike/work/llvm/12.0.0-rc2/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp, line 319.
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
Stack dump:
0.	Program arguments: llc _optimized.ll
1.	Running pass 'Function Pass Manager' on module '_optimized.ll'.
2.	Running pass 'X86 Assembly Printer' on function '@"test \22parse into struct with misc fields\22"'
Stack dump without symbol names (ensure you have llvm-symbolizer in your PATH or set the environment var `LLVM_SYMBOLIZER_PATH` to point to it):
0  llc                      0x0000000109fa01a7 llvm::sys::PrintStackTrace(llvm::raw_ostream&, int) + 39
1  llc                      0x0000000109f9ee98 llvm::sys::RunSignalHandlers() + 248
2  llc                      0x0000000109fa07f0 SignalHandler(int) + 272
3  libsystem_platform.dylib 0x00007fff203aed7d _sigtramp + 29
4  llc                      0x000000010d573569 llvm::DenseMapInfo<llvm::codeview::GloballyHashedType>::Tombstone + 248153
5  libsystem_c.dylib        0x00007fff202bd720 abort + 120
6  libsystem_c.dylib        0x00007fff202bc9d6 err + 0
7  llc                      0x000000010b034763 llvm::DbgVariable::addMMIEntry(llvm::DbgVariable const&) (.cold.5) + 35
8  llc                      0x000000010925128d llvm::DbgVariable::addMMIEntry(llvm::DbgVariable const&) + 525
9  llc                      0x0000000109258a1c llvm::DwarfDebug::collectVariableInfoFromMFTable(llvm::DwarfCompileUnit&, llvm::DenseSet<std::__1::pair<llvm::DINode const*, llvm::DILocation const*>, llvm::DenseMapInfo<std::__1::pair<llvm::DINode const*, llvm::DILocation const*> > >&) + 476
10 llc                      0x000000010925a5c2 llvm::DwarfDebug::collectEntityInfo(llvm::DwarfCompileUnit&, llvm::DISubprogram const*, llvm::DenseSet<std::__1::pair<llvm::DINode const*, llvm::DILocation const*>, llvm::DenseMapInfo<std::__1::pair<llvm::DINode const*, llvm::DILocation const*> > >&) + 66
11 llc                      0x000000010925bd39 llvm::DwarfDebug::endFunctionImpl(llvm::MachineFunction const*) + 265
12 llc                      0x000000010923d3b0 llvm::DebugHandlerBase::endFunction(llvm::MachineFunction const*) + 112
13 llc                      0x000000010922176d llvm::AsmPrinter::emitFunctionBody() + 13021
14 llc                      0x0000000108a2a36c llvm::X86AsmPrinter::runOnMachineFunction(llvm::MachineFunction&) + 300
15 llc                      0x000000010942f30d llvm::MachineFunctionPass::runOnFunction(llvm::Function&) + 349
16 llc                      0x0000000109817ea2 llvm::FPPassManager::runOnFunction(llvm::Function&) + 1202
17 llc                      0x000000010981e428 llvm::FPPassManager::runOnModule(llvm::Module&) + 72
18 llc                      0x000000010981851e llvm::legacy::PassManagerImpl::run(llvm::Module&) + 1102
19 llc                      0x0000000107d77feb main + 9723
20 libdyld.dylib            0x00007fff20385621 start + 1
```

**Comment by mikdusan** [comment]
@LemonBoy submitted a patch for the miscompilation https://reviews.llvm.org/D97665

**Comment by mikdusan** [comment]
looks like this combination of passes triggers the assert:

```sh
opt -inline -sroa -S _unoptimized.ll > _custom.ll
llc _custom.ll
```

**Comment by mikdusan** [comment]
files for this reduction: [reduction.zip](https://github.com/ziglang/zig/files/6075273/reduction.zip)

- new reduction
```sh
# NOTE: REDUCTION MUST BE RUN WITH ZIG BUILT AGAINST ONE OF:
# → LLVM with -DCMAKE_BUILD_TYPE=Release -DLLVM_ENABLE_ASSERTIONS=ON
# → LLVM with -DCMAKE_BUILD_TYPE=Debug

zig test reduction.zig -OReleaseFast
```
```
Assertion failed: ((FrameIndexExprs.size() == 1 || llvm::all_of(FrameIndexExprs, [](FrameIndexExpr &FIE) { return FIE.Expr && FIE.Expr->isFragment(); })) && "conflicting locations for variable"), function addMMIEntry, file /Users/mike/work/llvm/12.0.0-rc2/llvm/lib/CodeGen/AsmPrinter/DwarfDebug.cpp, line 319.
```

- `reduction.zig`
```zig
const std = @import("std");

test "" {
    const Foo = union(enum) {
        int: u32,
        float: f64,
        third: usize,
        fourth: isize,
        fifth: isize,
    };
    const rv = try std.json.parse(Foo, &std.json.TokenStream.init(""), .{});
}
```

- here's our big hint, grizzly bear with me:
- this `.ll` snippet correlates to multiple `@alloca` locals directly correlated to number of union members (five)
- look for 5 groups of locals named `tokens_copy`, `value` and `err` (**including** number suffixes to make unique)
- this is function `std.json.parseInternal`
```llvm
; Function Attrs: nobuiltin nounwind sspstrong
define internal fastcc void @std.json.parseInternal({ %Foo, i16 }* nonnull sret({ %Foo, i16 }) %0, %std.builtin.StackTrace* nonnull %1, %std.json.Token* nonnull readonly align 8 %2, %std.json.TokenStream* nonnull align 16 %3, %std.json.ParseOptions* nonnull readonly align 8 %4) unnamed_addr #1 !dbg !3592 {
Entry:
  %tokens_copy = alloca %std.json.TokenStream, align 16
  %5 = alloca %std.json.Token, align 8
  %6 = alloca %std.json.TokenStream*, align 8
  %7 = alloca %std.json.ParseOptions, align 8
  %8 = alloca { i32, i16 }, align 4
  %value = alloca i32, align 4
  %err = alloca i16, align 2
  %tokens_copy40 = alloca %std.json.TokenStream, align 16
  %9 = alloca %std.json.Token, align 8
  %10 = alloca %std.json.TokenStream*, align 8
  %11 = alloca %std.json.ParseOptions, align 8
  %12 = alloca { double, i16 }, align 8
  %value41 = alloca double, align 8
  %err42 = alloca i16, align 2
  %tokens_copy43 = alloca %std.json.TokenStream, align 16
  %13 = alloca %std.json.Token, align 8
  %14 = alloca %std.json.TokenStream*, align 8
  %15 = alloca %std.json.ParseOptions, align 8
  %16 = alloca { i64, i16 }, align 8
  %value44 = alloca i64, align 8
  %err45 = alloca i16, align 2
  %tokens_copy46 = alloca %std.json.TokenStream, align 16
  %17 = alloca %std.json.Token, align 8
  %18 = alloca %std.json.TokenStream*, align 8
  %19 = alloca %std.json.ParseOptions, align 8
  %20 = alloca { i64, i16 }, align 8
  %value47 = alloca i64, align 8
  %err48 = alloca i16, align 2
  %tokens_copy49 = alloca %std.json.TokenStream, align 16
  %21 = alloca %std.json.Token, align 8
  %22 = alloca %std.json.TokenStream*, align 8
  %23 = alloca %std.json.ParseOptions, align 8
  %24 = alloca { i64, i16 }, align 8
  %value50 = alloca i64, align 8
  %err51 = alloca i16, align 2
```

- now we get to what looks like the issue (thank you @LemonBoy for the big hint)
- for ref look at this switch block for locals: `tokens_copy`, `value`, `err`
 https://github.com/ziglang/zig/blob/0c06a1885fad9a9bb85342632a0b7c8a3a0733e9/lib/std/json.zig#L1510-L1532
- and finally in the same fn:
```
cat reduction.ll | grep @llvm.dbg.declare | grep tokens_copy
```
```llvm
  call void @llvm.dbg.declare(metadata %std.json.TokenStream* %tokens_copy, metadata !3605, metadata !DIExpression()), !dbg !3620
  call void @llvm.dbg.declare(metadata %std.json.TokenStream* %tokens_copy40, metadata !3605, metadata !DIExpression()), !dbg !3620
  call void @llvm.dbg.declare(metadata %std.json.TokenStream* %tokens_copy43, metadata !3605, metadata !DIExpression()), !dbg !3620
  call void @llvm.dbg.declare(metadata %std.json.TokenStream* %tokens_copy46, metadata !3605, metadata !DIExpression()), !dbg !3620
  call void @llvm.dbg.declare(metadata %std.json.TokenStream* %tokens_copy49, metadata !3605, metadata !DIExpression()), !dbg !3620
```

- notice we matched 5, the same number of union members in `reduction.zig`
- look to `metadata !NNNN` and see we have 5 debug decl pointing to the same variable:
```llvm
!3605 = !DILocalVariable(name: "tokens_copy", scope: !3606, file: !304, line: 1515, type: !2415)
```

In summary, it looks like something related to comptime inlining with some nested switch blocks leads to multiplication of locals, and those locals erroneously refer back to a single local for debug information purposes. I have thus far been unable to get a non-json-related reduction of this issue.

**Comment by andrewrk** [comment]
Confirmed that we're still hitting it with LLVM12rc4:

```
zig: ../lib/CodeGen/AsmPrinter/DwarfExpression.cpp:637: void llvm::DwarfExpression::addFragmentOffset(const llvm::DIExpression*): Assertion `FragmentOffset >= OffsetInBits && "overlapping or duplicate fragments"' failed.
The following command terminated unexpectedly:
/home/andy/dev/zig/build-llvm12-debug/zig test /home/andy/dev/zig/lib/std/std.zig --test-name-prefix std-native-ReleaseFast-bare-multi  -OReleaseFast --cache-dir /home/andy/dev/zig/zig-cache --global-cache-dir /home/andy/.cache/zig --name test -I /home/andy/dev/zig/test -L /home/andy/local/llvm12-debug/lib -isystem /home/andy/local/llvm12-debug/include --override-lib-dir /home/andy/dev/zig/lib 
```

**Comment by edt-xx** [comment]
Here the zls build works (#13745) using -Drelease-small with llvm 15.0.6 and zig 0.11.0-dev.491+e361740669, it fails with release-fast and release-safe.

**Comment by nektro** [comment]
upstream https://github.com/llvm/llvm-project/issues/33783

**Comment by alexrp** [comment]
I'm closing this because I just don't see an actual usable repro anywhere in this or the upstream issue, and I suspect this has been fixed since anyway.
